clippy_lints/src/hashset_insert_after_contains.rs

```diff
@@ -0,0 +1,137 @@
+use std::ops::ControlFlow;
+
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::ty::is_type_diagnostic_item;
+use clippy_utils::visitors::for_each_expr;
+use clippy_utils::{higher, peel_hir_expr_while, SpanlessEq};
+use rustc_hir::{Expr, ExprKind, UnOp};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::{sym, Span};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `contains` to see if a value is not
+    /// present on `HashSet` followed by a `insert`.
+    ///
+    /// ### Why is this bad?
+    /// Using just `insert` and checking the returned `bool` is more efficient.
+    ///
```

Comment:
I know of one possible false positive: If the value is only borrowed & expensive to clone or impossible to clone twice, we may opt to check with `contains` before inserting to avoid the clone. There should be a "known problems" section mentioning this.

Reply:
Ok, did it here https://github.com/rust-lang/rust-clippy/pull/12873/commits/2b0cad6451cfe192e251379b741e1214cb36623a